### ISSUE TITLE
Estimator Training Data Reading Incorrect Values

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [tensorflow/tensorflow#56476]
### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.5+

### Custom Code

No

### OS Platform and Distribution

Mac and Linux

### Python version

3.8

### Current Behaviour?

Starting with tensorflow 2.5, I've encountered my training pipeline having wrong values for certain intermediate calculation. It looks like the data pipeline is reading wrong value in tf.train.Example. I made toy dataset with 1 feature (all -5) and 1 label (always 1) and I print the label, but instead it shows value of the feature.

I've been able to reproduce it down to small estimator training loop although it's still ~50 lines. The model here is trivial fake model as focus is on the value of label tensor passed to model_fn. Dataset pipeline is short but still several steps. If I remove some steps the bug disappears. Also batching uses,

`dataset = dataset.batch(batch_size, drop_remainder=True)`

The drop_remainder is required for the bug. Changing that somehow changes how the data is read.

In the model_fn, minor what should be equivalent variations of this line effect whether bug happens.

`tf.stack([features[key] for key in label_names])`

It's also not specific only to tf.stack. tf.reduce_mean also leads to negative values even though features[key] is always positive. One key element is that this list has same tensor multiple times because label is repeated across different model heads. If label_names is unique the bug disappears.

I've tested this on 2.4, 2.5, 2.6, 2.7, 2.8, 2.9. All of 2.5+ consistently show same wrong behavior for value of intermediate printed. I am working on bisecting to exact commit, but building tensorflow is slow. So far I've narrowed it down to bug being introduced in between March 18 commit [fd020fff99a0](https://github.com/tensorflow/tensorflow/commit/fd020fff99a0ad5e0293e43a77df54dd9efedce7) and March 24 commit [83a8259081](https://github.com/tensorflow/tensorflow/commits/b9e31e669e454b185a100802b5d6cb1a6f59d74b?before=b9e31e669e454b185a100802b5d6cb1a6f59d74b+140&branch=b9e31e669e454b185a100802b5d6cb1a6f59d74b&qualified_name=b9e31e669e454b185a100802b5d6cb1a6f59d74b).

I've also tested on both my laptop (mac) and a linux gcp VM. All tests were done with CPU only. They show same behavior.

My suspicion is some memory aliasing is happening when non unique tensor appears for some of these operations. I'm confused why this is estimator specific. Maybe it's possible to reproduce this without estimator, but I've been unsuccessful with that.

### Standalone code to reproduce the issue

Here's a colab link, https://colab.research.google.com/drive/11-7vyU4cc_3BGXudUEuT12bbVS-JJQPH?usp=sharing It is fully self contained with only dependency being tensorflow. If you run notebook in different tensorflow versions <=2.4 vs >=2.5 it produces different behavior.

### Relevant log output

If you run notebook on a good tensorflow version (2.4) you will see 1 for label. If you run it on any bad version (2.5+) -5 is result.

**Comment by tilakrayal** [comment]
@sachinprasadhs,
I was able to reproduce the issue on tensorflow [v2.8](https://colab.research.google.com/gist/tilakrayal/d23df539a6010d8967c72be153871509/2-8-estimator_training_data_misreading.ipynb), [v2.9](https://colab.research.google.com/gist/tilakrayal/cfffcca56cf99d82006c575e27558016/2-9estimator_training_data_misreading.ipynb) and nightly. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/4bf30e92808128eb05c0851468d87385/2-10-estimator_training_data_misreading.ipynb).

**Comment by sachinprasadhs** [comment]
Hi, Estimator will be deprecated from Tensorflow 2.12 since it uses Tensorflow V1 behavior.
Could you please migrate from Estimator to Keras API which will enable you to perform the aforementioned tasks, such as [model building](https://www.tensorflow.org/guide/keras/custom_layers_and_models), gradient application, [training](https://www.tensorflow.org/guide/keras/customizing_what_happens_in_fit), evaluation, and prediction.
Refer Migration guide here https://www.tensorflow.org/guide/migrate/migrating_estimator for more details. Thanks!
